### PR TITLE
Adopt pure ruby DSL implementation for JRuby

### PR DESCRIPTION
**Pull request by JasonLunn**
JRuby unit and conformance test fixes.

Includes changes originally submitted in #8997 and #8954

### CODE REVIEW

File: ruby/pom.xml [hunk 1]
```diff
@@ -88,7 +88,7 @@
         <dependency>
             <groupId>com.google.protobuf</groupId>
             <artifactId>protobuf-java-util</artifactId>
-            <version>3.13.0</version>
+            <version>3.18.0</version>
```
Comment (perezd) [inline ruby/pom.xml:91]: I think you'll need to add an entry to this:
https://github.com/protocolbuffers/protobuf/blob/master/update_version.py

to ensure this stays updated.

**Review by perezd — APPROVED** [review]
LGTM!

File: ruby/pom.xml (hunk 1 above)
Comment (JasonLunn) [reply ruby/pom.xml]: Done - thanks for the pointer.

File: ruby/Rakefile [hunk 2]
```diff
@@ -67,16 +67,18 @@ unless ENV['IN_DOCKER'] == 'true'
 end
 
 if RUBY_PLATFORM == "java"
-  if `which mvn` == ''
-    raise ArgumentError, "maven needs to be installed"
-  end
-  task :clean do
+  task :clean => :require_mvn do
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/Rakefile:70]: The JRuby gem assembly process has a dependency on Maven - they just don't need to be enforced unless you're actually running one of the JRuby tasks; previously it was enforcing the availability of Maven when the tasks were defined.

File: ruby/lib/google/protobuf.rb [hunk 3]
```diff
@@ -51,9 +51,10 @@ class TypeError < ::TypeError; end
     require 'google/protobuf_c'
   end
 
-  require 'google/protobuf/descriptor_dsl'
 end
 
+require 'google/protobuf/descriptor_dsl'
+
```
[the remark this answers is not in the document]
Comment (JasonLunn) [reply ruby/lib/google/protobuf.rb]: Done

**Review by haberman — APPROVED** [review]

**Pull request merged**